1. If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.


In [1]:
sum = 0
multiples = []
for i in range(1, 1000):
    if i % 3 == 0 or i % 5 == 0:
        multiples.append(i)
        sum = sum + i
print('Multiples of 3 or 5 below 1000:\n', multiples)
print('\nSum of all the multiples of 3 or 5 below 1000:\n', sum)

Multiples of 3 or 5 below 1000:
 [3, 5, 6, 9, 10, 12, 15, 18, 20, 21, 24, 25, 27, 30, 33, 35, 36, 39, 40, 42, 45, 48, 50, 51, 54, 55, 57, 60, 63, 65, 66, 69, 70, 72, 75, 78, 80, 81, 84, 85, 87, 90, 93, 95, 96, 99, 100, 102, 105, 108, 110, 111, 114, 115, 117, 120, 123, 125, 126, 129, 130, 132, 135, 138, 140, 141, 144, 145, 147, 150, 153, 155, 156, 159, 160, 162, 165, 168, 170, 171, 174, 175, 177, 180, 183, 185, 186, 189, 190, 192, 195, 198, 200, 201, 204, 205, 207, 210, 213, 215, 216, 219, 220, 222, 225, 228, 230, 231, 234, 235, 237, 240, 243, 245, 246, 249, 250, 252, 255, 258, 260, 261, 264, 265, 267, 270, 273, 275, 276, 279, 280, 282, 285, 288, 290, 291, 294, 295, 297, 300, 303, 305, 306, 309, 310, 312, 315, 318, 320, 321, 324, 325, 327, 330, 333, 335, 336, 339, 340, 342, 345, 348, 350, 351, 354, 355, 357, 360, 363, 365, 366, 369, 370, 372, 375, 378, 380, 381, 384, 385, 387, 390, 393, 395, 396, 399, 400, 402, 405, 408, 410, 411, 414, 415, 417, 420, 423, 425, 426, 429, 430, 432, 435, 4

2. Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.


In [5]:
sum = 0
previousNumber = 0
newNumber = 1
maximumValue = 4000000
i = 1
print('Terms in the Fibonacci sequence that do not exceed four million and are even-valued:')

while newNumber < maximumValue:
    aux = previousNumber
    previousNumber = newNumber
    newNumber = aux + newNumber
    if newNumber > maximumValue:
        break
    if newNumber % 2 == 0:
        print(newNumber)
        sum = sum + newNumber
print('\nSum of the even-valued terms:',sum)


Terms in the Fibonacci sequence that do not exceed four million and are even-valued:
2
8
34
144
610
2584
10946
46368
196418
832040
3524578

Sum of the even-valued terms: 4613732


3. The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?


In [65]:
def maxValue(listOfElements):
    largestElement = 0
    for i in listOfElements:
        if i > largestElement:
            largestElement = i
    return largestElement
# number = int(input('Enter a number: '))
# number = 13195
number = 600851475143
print('Entered number:',number)
aux = 2
listOfPrimeFactors = []
while aux * aux <= number:
    if number % aux:
        aux = aux + 1
        # print('aux =',aux)
        # print('number',number)
        # print()
    else:
        number = number // aux
        # print('aux =',aux)
        listOfPrimeFactors.append(aux)
if number > 1:
    listOfPrimeFactors.append(number)
print('List of prime factors:',listOfPrimeFactors)
# largestPrimeFactor = 0
# for i in listOfPrimeFactors:
#     if i > largestPrimeFactor:
#         largestPrimeFactor = i
# print('The largest prime factor of the given number:',largestPrimeFactor)
print('The largest prime factor of the given number:',maxValue(listOfPrimeFactors))


Entered number: 600851475143
List of prime factors: [71, 839, 1471, 6857]
The largest prime factor of the given number: 6857


4. A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.


In [90]:
def verifyPalindrome(product):
    palindrome = 0
    aux = product
    while True:
        # print('product =', product)
        if product == 0:
            return False
        lastDigit = product % 10
        product //= 10
        # print(lastDigit)
        palindrome *= 10
        palindrome += lastDigit
        # print('palindrome =',palindrome)
        # print()
        if aux == palindrome:
            return True
flag = None
# palindromes = []
palindromes_dict = {}
for i in range(999, 99, -1):
    for j in range(999, 99, -1):
        # print('i:',i,', j:',j)
        product = i*j
        # print(str(i) + ' * ' + str(j) + ' =',product)
        if verifyPalindrome(product):
            # print('Palindrome product:',product)
            # flag = True
            # break
            # palindromes.append(product)
            palindromes_dict[str(i)+' * '+str(j)] = product
            product = 0
    # if flag:
    #     break
# print('List of palindromes:')
# for p in palindromes:
#     print(p)
print('\nDictionary of palindromes:')
for item in palindromes_dict.items():
    print(item)
# print('\nLargest palindrome made from the product of two 3-digit numbers:',maxValue(palindromes))
print('\nLargest palindrome made from the product of two 3-digit numbers:',maxValue(palindromes_dict.values()))
print()
print(type(palindromes_dict.values()))
m = maxValue(palindromes_dict.values())
# index = list(palindromes_dict.values()).index(m)
print('m =',m)
# print('index =', index)
for item in palindromes_dict.items():
    # print(type(item))
    # break
    # print(item[1])
    if item[1] == m:
        print(item)
        break

# print('\nLargest palindrome made from the product of two 3-digit numbers:',list(palindromes_dict.keys()).index(maxValue(palindromes_dict.values())))


Dictionary of palindromes:
('995 * 583', 580085)
('995 * 517', 514415)
('993 * 913', 906609)
('991 * 121', 119911)
('987 * 286', 282282)
('987 * 143', 141141)
('982 * 869', 853358)
('979 * 664', 650056)
('979 * 614', 601106)
('979 * 605', 592295)
('979 * 555', 543345)
('979 * 496', 485584)
('979 * 446', 436634)
('979 * 387', 378873)
('979 * 337', 329923)
('978 * 418', 408804)
('978 * 209', 204402)
('977 * 638', 623326)
('975 * 539', 525525)
('974 * 308', 299992)
('971 * 484', 469964)
('968 * 916', 886688)
('968 * 831', 804408)
('968 * 722', 698896)
('968 * 652', 631136)
('968 * 637', 616616)
('968 * 458', 443344)
('968 * 443', 428824)
('968 * 279', 270072)
('968 * 264', 255552)
('966 * 858', 828828)
('966 * 429', 414414)
('964 * 506', 487784)
('963 * 187', 180081)
('962 * 924', 888888)
('962 * 693', 666666)
('962 * 462', 444444)
('962 * 231', 222222)
('961 * 363', 348843)
('959 * 803', 770077)
('958 * 682', 653356)
('957 * 894', 855558)
('957 * 844', 807708)
('957 * 728', 696696)
('95

5. 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?


In [163]:
# def verifyIfPrime(number):
#     if number == 0 or number == 1:
#         return False
#     for i in range(2, number):
#         if number % i == 0:
#             return False
#     return True
#
# listOfNumbers = []
# listOfPrimeNumbers = []
# for i in range(1,21):
#     listOfNumbers.append(i)
#     if verifyIfPrime(i):
#         listOfPrimeNumbers.append(i)
# print('List of numbers:',listOfNumbers)
# print('List of prime numbers:',listOfPrimeNumbers)
# --------------------------------------------------

def getPrimeFactors(number):
    aux = 2
    listOfPrimeFactors = []
    while aux * aux <= number:
        if number % aux:
            aux = aux + 1
        else:
            number = number // aux
            listOfPrimeFactors.append(aux)
    if number > 1:
        listOfPrimeFactors.append(number)
    return listOfPrimeFactors

# def getLargestPrimeFactor(number):
#     aux = 2
#     listOfPrimeFactors = []
#     while aux * aux <= number:
#         if number % aux:
#             aux = aux + 1
#         else:
#             number = number // aux
#             listOfPrimeFactors.append(aux)
#     if number > 1:
#         listOfPrimeFactors.append(number)
#     return maxValue(listOfPrimeFactors)

# number = int(input('Enter a number: '))
def lcm(listOfNumbers):
    # largestPrimeFactorList = []
    # for number in listOfNumbers:
    #     largestPrimeFactorList.append(getLargestPrimeFactor(number))
    # print(largestPrimeFactorList)
    primeFactorsLCM = []
    for number in listOfNumbers:
        primeFactorsLCM.append(getPrimeFactors(number))
    # print(primeFactorsLCM)

    powersDict = {}
    for power in primeFactorsLCM:
        # print(power)
        maxPower = maxValue(power)
        powersDict[maxPower] = power
    # print(list(powersDict.items()))
    # listOfHighestPoweredFactors = []
    productOfHighestPoweredFactor = 1
    for pair in powersDict.items():
        # print(list(pair))
        for value in pair[1]:
            if value == pair[0]:
                productOfHighestPoweredFactor*=value
                # listOfHighestPoweredFactors.append(productOfHighestPoweredFactor)
    # print(productOfHighestPoweredFactor)
    return productOfHighestPoweredFactor

# For verification at online calculator:
# https://www.calculatorsoup.com/calculators/math/lcm.php
# 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
print(lcm([number for number in range(1,21)]))

232792560


6. The sum of the squares of the first ten natural numbers is, 1^2 + 2^2 + ... + 10^2 = 385

The square of the sum of the first ten natural numbers is, (1 + 2 + ... + 10)^2 = 3025

Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is, 3025 - 385 = 2640

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [181]:
listOfSquares = [value**2 for value in range(1,101)]
print(listOfSquares)
sumOfTheSquares = 0
for number in listOfSquares:
    sumOfTheSquares+=number
print('Sum of the squares:',sumOfTheSquares)
print()
sumOfNaturalNumbers = 0
squareOfTheSum = 0
listOfNaturalNumbers = [value for value in range(1,101)]
print(listOfNaturalNumbers)
for number in listOfNaturalNumbers:
    sumOfNaturalNumbers += number
squareOfTheSum = sumOfNaturalNumbers**2
print('Square of the sum:',squareOfTheSum)
print('\nDifference:',squareOfTheSum-sumOfTheSquares)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801, 10000]
Sum of the squares: 338350

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100]
Square of the sum: 2

7. By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10001st prime number?


In [1]:
# import math

def verifyIfPrime(number):
    if number == 0 or number == 1:
        return False
    for i in range(2, number):
        if number % i == 0:
            return False
    return True

n = int(input('Number of prime numbers wanted:'))
# listOfNumbers = []
listOfPrimeNumbers = []
i = 2
while True:
    # listOfNumbers.append(i)
    if len(listOfPrimeNumbers) < n:
        if verifyIfPrime(i):
            listOfPrimeNumbers.append(i)
        i += 1
    elif len(listOfPrimeNumbers) >= n:
        break
# print('List of numbers:',listOfNumbers)
print('List of prime numbers:',listOfPrimeNumbers)

# By using the Sieve of Eratosthenes:
# def sieveForPrimes(n):
#     largerNumberOfPrimes = int(2 * n * math.log(n))
#     sieve = [True] * largerNumberOfPrimes
#     count = 0
#     for i in range(2, largerNumberOfPrimes):
#         if sieve[i]:
#             count += 1
#             if count == n:
#                 return i
#             for j in range(2*i, largerNumberOfPrimes, i):
#                 sieve[j] = False
# sieveForPrimes(10001)

104743

8. A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
a^2 + b^2 = c^2

For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.


In [6]:
def pythagoreanTripletProduct():
    for c in range(2, 1001):
        for a in range(1, c+1):
            b = 1000 - c - a
            if a**2 + b**2 == c**2:
                print('a =',a)
                print('b =',b)
                print('c =',c)
                print('a*b*c =', a*b*c)
                return

pythagoreanTripletProduct()

a = 200
b = 375
c = 425
a*b*c = 31875000


9. The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.


In [2]:
initial_sum=17
for i in range(11,2000000):
    for j in range(2,int(i**0.5+1)):
        if i%j==0:
            break
    else:
        initial_sum+=i
print(initial_sum)

142913828922


10. The sequence of triangle numbers is generated by adding the natural numbers. So the 7th triangle number would be 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28. The first ten terms would be:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

Let us list the factors of the first seven triangle numbers:

     1: 1
     3: 1,3
     6: 1,2,3,6
    10: 1,2,5,10
    15: 1,3,5,15
    21: 1,3,7,21
    28: 1,2,4,7,14,28

We can see that 28 is the first triangle number to have over five divisors.

What is the value of the first triangle number to have over five hundred divisors?


In [22]:
def divisors(n):
    numberOfFactors = 0
    for i in range(1, int(math.ceil(math.sqrt(n)))+2):
        if n % i == 0:
            numberOfFactors += 2
        if i * i == n:
            numberOfFactors -= 1
    return numberOfFactors

for n in range(1,1000001):
    tr = (n*(n+1))/2
    print('tr =',tr)
    if n % 2 == 0:
        count = divisors(n/2) * divisors(n+1)
    else:
        count = divisors(n) * divisors((n+1)/2)
    if count >= 500:
        print(tr)
        break

tr = 1.0
tr = 3.0
tr = 6.0
tr = 10.0
tr = 15.0
tr = 21.0
tr = 28.0
tr = 36.0
tr = 45.0
tr = 55.0
tr = 66.0
tr = 78.0
tr = 91.0
tr = 105.0
tr = 120.0
tr = 136.0
tr = 153.0
tr = 171.0
tr = 190.0
tr = 210.0
tr = 231.0
tr = 253.0
tr = 276.0
tr = 300.0
tr = 325.0
tr = 351.0
tr = 378.0
tr = 406.0
tr = 435.0
tr = 465.0
tr = 496.0
tr = 528.0
tr = 561.0
tr = 595.0
tr = 630.0
tr = 666.0
tr = 703.0
tr = 741.0
tr = 780.0
tr = 820.0
tr = 861.0
tr = 903.0
tr = 946.0
tr = 990.0
tr = 1035.0
tr = 1081.0
tr = 1128.0
tr = 1176.0
tr = 1225.0
tr = 1275.0
tr = 1326.0
tr = 1378.0
tr = 1431.0
tr = 1485.0
tr = 1540.0
tr = 1596.0
tr = 1653.0
tr = 1711.0
tr = 1770.0
tr = 1830.0
tr = 1891.0
tr = 1953.0
tr = 2016.0
tr = 2080.0
tr = 2145.0
tr = 2211.0
tr = 2278.0
tr = 2346.0
tr = 2415.0
tr = 2485.0
tr = 2556.0
tr = 2628.0
tr = 2701.0
tr = 2775.0
tr = 2850.0
tr = 2926.0
tr = 3003.0
tr = 3081.0
tr = 3160.0
tr = 3240.0
tr = 3321.0
tr = 3403.0
tr = 3486.0
tr = 3570.0
tr = 3655.0
tr = 3741.0
tr = 3828.0
tr = 3916.0
tr =